File: packages/python/anip-server/tests/compliance.py
"""Backend compliance test suite for StorageBackend implementations."""
import asyncio
from typing import Any


async def compliance_token_roundtrip(storage) -> None:
    """Token store/load roundtrip."""
    token = {"token_id": "tok-c1", "issuer": "svc", "subject": "agent",
             "scope": ["a.b"], "purpose": {"capability": "c", "parameters": {}, "task_id": "t"},
             "parent": None, "expires": "2099-01-01T00:00:00Z",
             "constraints": {"max_delegation_depth": 3, "concurrent_branches": "allowed"},
             "root_principal": "human:alice"}
    await storage.store_token(token)
    loaded = await storage.load_token("tok-c1")
    assert loaded is not None
    assert loaded["token_id"] == "tok-c1"
    assert loaded["scope"] == ["a.b"]


async def compliance_token_not_found(storage) -> None:
    """load_token returns None for unknown ID."""
    assert await storage.load_token("nonexistent") is None


async def compliance_audit_roundtrip(storage) -> None:
    """Audit entry store and query."""
    entry = {"sequence_number": 1, "timestamp": "2026-01-01T00:00:00Z",
             "capability": "search", "token_id": "tok-1",
             "root_principal": "human:alice", "success": True,
             "invocation_id": "inv-aabbccddeeff", "client_reference_id": "ref-1",
             "previous_hash": "sha256:0", "signature": None}
    await storage.store_audit_entry(entry)
    results = await storage.query_audit_entries(capability="search")
    assert len(results) >= 1
    assert results[0]["capability"] == "search"


async def compliance_audit_lineage_filters(storage) -> None:
    """Audit query by invocation_id and client_reference_id."""
    for i in range(3):
        await storage.store_audit_entry({
            "sequence_number": i + 1, "timestamp": f"2026-01-0{i+1}T00:00:00Z",
            "capability": "cap", "root_principal": "human:a", "success": True,
            "invocation_id": f"inv-{'0' * 11}{i}", "client_reference_id": "ref-shared",
            "previous_hash": "sha256:0", "signature": None})
    by_inv = await storage.query_audit_entries(invocation_id="inv-000000000001")
    assert len(by_inv) == 1
    by_ref = await storage.query_audit_entries(client_reference_id="ref-shared")
    assert len(by_ref) == 3


async def compliance_audit_ordering(storage) -> None:
    """Audit entries maintain insertion order by sequence number."""
    for i in range(5):
        await storage.store_audit_entry({
            "sequence_number": i + 1, "timestamp": f"2026-01-0{i+1}T00:00:00Z",
            "capability": "cap", "root_principal": "human:a", "success": True,
            "previous_hash": "sha256:0", "signature": None})
    entries = await storage.query_audit_entries(root_principal="human:a", limit=10)
    seq_nums = [e["sequence_number"] for e in entries]
    # Descending order (most recent first)
    assert seq_nums == sorted(seq_nums, reverse=True)


async def compliance_audit_concurrent_ordering(storage) -> None:
    """Audit insertion ordering under concurrent calls."""
    async def insert(seq: int):
        await storage.store_audit_entry({
            "sequence_number": seq, "timestamp": "2026-01-01T00:00:00Z",
            "capability": "cap", "root_principal": "human:a", "success": True,
            "previous_hash": "sha256:0", "signature": None})

    await asyncio.gather(*[insert(i + 1) for i in range(20)])
    entries = await storage.query_audit_entries(root_principal="human:a", limit=100)
    assert len(entries) == 20
    seq_nums = [e["sequence_number"] for e in entries]
    assert seq_nums == sorted(seq_nums, reverse=True)
    last = await storage.get_last_audit_entry()
    assert last is not None
    assert last["sequence_number"] == 20


async def compliance_checkpoint_roundtrip(storage) -> None:
    """Checkpoint store/load roundtrip."""
    body = {"checkpoint_id": "cp-1", "merkle_root": "sha256:abc",
            "range": {"first_sequence": 1, "last_sequence": 5},
            "timestamp": "2026-01-01T00:00:00Z", "entry_count": 5}
    await storage.store_checkpoint(body, "sig-123")
    loaded = await storage.get_checkpoint_by_id("cp-1")
    assert loaded is not None
    assert loaded["checkpoint_id"] == "cp-1"
    assert loaded["merkle_root"] == "sha256:abc"


async def compliance_checkpoint_not_found(storage) -> None:
    """get_checkpoint_by_id returns None for unknown ID."""
    assert await storage.get_checkpoint_by_id("nonexistent") is None


async def compliance_checkpoint_listing(storage) -> None:
    """Checkpoint listing respects limit."""
    for i in range(5):
        await storage.store_checkpoint(
            {"checkpoint_id": f"cp-{i}", "merkle_root": f"sha256:{i}", "sequence_number": i + 1},
            f"sig-{i}")
    results = await storage.get_checkpoints(limit=3)
    assert len(results) == 3


async def compliance_audit_entries_range(storage) -> None:
    """get_audit_entries_range returns entries between sequence numbers."""
    for i in range(10):
        await storage.store_audit_entry({
            "sequence_number": i + 1, "timestamp": f"2026-01-{i+1:02d}T00:00:00Z",
            "capability": "cap", "root_principal": "human:a", "success": True,
            "previous_hash": "sha256:0", "signature": None})
    entries = await storage.get_audit_entries_range(3, 7)
    seq_nums = [e["sequence_number"] for e in entries]
    assert all(3 <= s <= 7 for s in seq_nums)


# Collected list for parametrize
ALL_COMPLIANCE_TESTS = [
    compliance_token_roundtrip,
    compliance_token_not_found,
    compliance_audit_roundtrip,
    compliance_audit_lineage_filters,
    compliance_audit_ordering,
    compliance_audit_concurrent_ordering,
    compliance_checkpoint_roundtrip,
    compliance_checkpoint_not_found,
    compliance_checkpoint_listing,
    compliance_audit_entries_range,
]

File: packages/typescript/server/tests/compliance.ts
/**
 * Backend compliance test suite for StorageBackend implementations.
 *
 * Each fixture is an async function that accepts a {@link StorageBackend} and
 * exercises one aspect of the contract.  The {@link ALL_COMPLIANCE_TESTS}
 * array collects them so that test runners can iterate over them for every
 * backend variant (InMemoryStorage, SQLiteStorage, etc.).
 */

import { expect } from "vitest";
import type { StorageBackend } from "../src/storage.js";

// ---------------------------------------------------------------------------
// Token fixtures
// ---------------------------------------------------------------------------

export async function complianceTokenRoundtrip(storage: StorageBackend): Promise<void> {
  const token = {
    token_id: "tok-c1",
    issuer: "svc",
    subject: "agent",
    scope: ["a.b"],
    purpose: { capability: "c", parameters: {}, task_id: "t" },
    parent: null,
    expires: "2099-01-01T00:00:00Z",
    constraints: { max_delegation_depth: 3, concurrent_branches: "allowed" },
    root_principal: "human:alice",
  };
  await storage.storeToken(token);
  const loaded = await storage.loadToken("tok-c1");
  expect(loaded).not.toBeNull();
  expect(loaded!.token_id).toBe("tok-c1");
  expect(loaded!.scope).toEqual(["a.b"]);
}

export async function complianceTokenNotFound(storage: StorageBackend): Promise<void> {
  expect(await storage.loadToken("nonexistent")).toBeNull();
}

// ---------------------------------------------------------------------------
// Audit fixtures
// ---------------------------------------------------------------------------

export async function complianceAuditRoundtrip(storage: StorageBackend): Promise<void> {
  const entry = {
    sequence_number: 1,
    timestamp: "2026-01-01T00:00:00Z",
    capability: "search",
    token_id: "tok-1",
    root_principal: "human:alice",
    success: true,
    invocation_id: "inv-aabbccddeeff",
    client_reference_id: "ref-1",
    previous_hash: "sha256:0",
    signature: null,
  };
  await storage.storeAuditEntry(entry);
  const results = await storage.queryAuditEntries({ capability: "search" });
  expect(results.length).toBeGreaterThanOrEqual(1);
  expect(results[0].capability).toBe("search");
}

export async function complianceAuditLineageFilters(storage: StorageBackend): Promise<void> {
  for (let i = 0; i < 3; i++) {
    await storage.storeAuditEntry({
      sequence_number: i + 1,
      timestamp: `2026-01-0${i + 1}T00:00:00Z`,
      capability: "cap",
      root_principal: "human:a",
      success: true,
      invocation_id: `inv-${"0".repeat(11)}${i}`,
      client_reference_id: "ref-shared",
      previous_hash: "sha256:0",
      signature: null,
    });
  }
  const byInv = await storage.queryAuditEntries({ invocationId: "inv-000000000001" });
  expect(byInv).toHaveLength(1);
  const byRef = await storage.queryAuditEntries({ clientReferenceId: "ref-shared" });
  expect(byRef).toHaveLength(3);
}

export async function complianceAuditOrdering(storage: StorageBackend): Promise<void> {
  for (let i = 0; i < 5; i++) {
    await storage.storeAuditEntry({
      sequence_number: i + 1,
      timestamp: `2026-01-0${i + 1}T00:00:00Z`,
      capability: "cap",
      root_principal: "human:a",
      success: true,
      previous_hash: "sha256:0",
      signature: null,
    });
  }
  const entries = await storage.queryAuditEntries({ rootPrincipal: "human:a", limit: 10 });
  const seqNums = entries.map((e) => e.sequence_number as number);
  // Descending order (most recent first)
  expect(seqNums).toEqual([...seqNums].sort((a, b) => b - a));
}

export async function complianceAuditConcurrentOrdering(storage: StorageBackend): Promise<void> {
  const insert = async (seq: number) => {
    await storage.storeAuditEntry({
      sequence_number: seq,
      timestamp: "2026-01-01T00:00:00Z",
      capability: "cap",
      root_principal: "human:a",
      success: true,
      previous_hash: "sha256:0",
      signature: null,
    });
  };

  await Promise.all(Array.from({ length: 20 }, (_, i) => insert(i + 1)));
  const entries = await storage.queryAuditEntries({ rootPrincipal: "human:a", limit: 100 });
  expect(entries).toHaveLength(20);
  const seqNums = entries.map((e) => e.sequence_number as number);
  expect(seqNums).toEqual([...seqNums].sort((a, b) => b - a));
  const last = await storage.getLastAuditEntry();
  expect(last).not.toBeNull();
  expect(last!.sequence_number).toBe(20);
}

// ---------------------------------------------------------------------------
// Checkpoint fixtures
// ---------------------------------------------------------------------------

export async function complianceCheckpointRoundtrip(storage: StorageBackend): Promise<void> {
  const body = {
    checkpoint_id: "cp-1",
    merkle_root: "sha256:abc",
    range: { first_sequence: 1, last_sequence: 5 },
    timestamp: "2026-01-01T00:00:00Z",
    entry_count: 5,
  };
  await storage.storeCheckpoint(body, "sig-123");
  const loaded = await storage.getCheckpointById("cp-1");
  expect(loaded).not.toBeNull();
  expect(loaded!.checkpoint_id).toBe("cp-1");
  expect(loaded!.merkle_root).toBe("sha256:abc");
}

export async function complianceCheckpointNotFound(storage: StorageBackend): Promise<void> {
  expect(await storage.getCheckpointById("nonexistent")).toBeNull();
}

export async function complianceCheckpointListing(storage: StorageBackend): Promise<void> {
  for (let i = 0; i < 5; i++) {
    await storage.storeCheckpoint(
      { checkpoint_id: `cp-${i}`, merkle_root: `sha256:${i}`, sequence_number: i + 1 },
      `sig-${i}`,
    );
  }
  const results = await storage.getCheckpoints(3);
  expect(results).toHaveLength(3);
}

// ---------------------------------------------------------------------------
// Audit range fixture
// ---------------------------------------------------------------------------

export async function complianceAuditEntriesRange(storage: StorageBackend): Promise<void> {
  for (let i = 0; i < 10; i++) {
    await storage.storeAuditEntry({
      sequence_number: i + 1,
      timestamp: `2026-01-${String(i + 1).padStart(2, "0")}T00:00:00Z`,
      capability: "cap",
      root_principal: "human:a",
      success: true,
      previous_hash: "sha256:0",
      signature: null,
    });
  }
  const entries = await storage.getAuditEntriesRange(3, 7);
  const seqNums = entries.map((e) => e.sequence_number as number);
  expect(seqNums.every((s) => s >= 3 && s <= 7)).toBe(true);
}

// ---------------------------------------------------------------------------
// Collected list
// ---------------------------------------------------------------------------

export const ALL_COMPLIANCE_TESTS = [
  { name: "token roundtrip", fn: complianceTokenRoundtrip },
  { name: "token not found", fn: complianceTokenNotFound },
  { name: "audit roundtrip", fn: complianceAuditRoundtrip },
  { name: "audit lineage filters", fn: complianceAuditLineageFilters },
  { name: "audit ordering", fn: complianceAuditOrdering },
  { name: "audit concurrent ordering", fn: complianceAuditConcurrentOrdering },
  { name: "checkpoint roundtrip", fn: complianceCheckpointRoundtrip },
  { name: "checkpoint not found", fn: complianceCheckpointNotFound },
  { name: "checkpoint listing", fn: complianceCheckpointListing },
  { name: "audit entries range", fn: complianceAuditEntriesRange },
];
